### PR TITLE
CDAP-17903 Allow reusing connection in pipeline

### PR DESCRIPTION
**Pull request by yaojiefeng**
Allow reusing connection in pipeline, 
this pr basically does: 
1. Add a children field if a config contains a collection of configs
2. Add deploy plugin and instantiate logic based on the above changes
3. Add ConnectionMacroEvaluator to evaluate macros at runtime

### CODE REVIEW

File: cdap-api/src/main/java/io/cdap/cdap/api/plugin/PluginGroupConfig.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import io.cdap.cdap.api.Config;
+
+/**
+ * Config that can be defined as a field inside {@link PluginConfig} to represent a collection of configs.
+ * When the plugin is deployed, the configs inside this class will be inspected and represented in the same way as
+ * {@link PluginPropertyField}.
+ * The {@link PluginPropertyField} for this field will contain a collection of property names about configs inside
+ * this class.
+ */
+public class PluginGroupConfig extends Config {
```
Comment (albertshau) [inline cdap-api/src/main/java/io/cdap/cdap/api/plugin/PluginGroupConfig.java]: Config is an application config and shouldn't be extended.

I'm wondering whether we need this class, could the contract be that the field should just extend the existing PluginConfig class? Wouldn't we need some of PluginConfig's methods for checking if a property contains a macro anyway?

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/common/AbstractServiceRetryableMacroEvaluator.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+  @Override
+  public String evaluate(String macroFunction, String... args) throws InvalidMacroException {
+    if (!functionName.equals(macroFunction)) {
+      // This shouldn't happen
+      throw new IllegalArgumentException("Invalid function name " + macroFunction
+                                           + ". Expecting " + functionName);
+    }
+
+    // Make call to get OAuth token with exponential delay on failure retry.
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/common/AbstractServiceRetryableMacroEvaluator.java]: comment looks wrong

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/common/ConnectionMacroEvaluator.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  String evaluateMacro(String macroFunction,
+                       String... args) throws InvalidMacroException, IOException, RetryableException {
+    if (args.length != 1) {
+      throw new InvalidMacroException("Macro '" + FUNCTION_NAME + "' should have exactly 1 arguments");
+    }
+
+    HttpURLConnection urlConn = serviceDiscoverer.openConnection(NamespaceId.SYSTEM.getNamespace(),
+                                                                 Constants.PIPELINEID,
+                                                                 Constants.STUDIO_SERVICE_NAME,
+                                                                 String.format("v1//contexts/%s/connections/%s",
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/common/ConnectionMacroEvaluator.java]: should be '/' instead of '//'?

File: cdap-api-common/src/main/java/io/cdap/cdap/internal/io/SchemaTypeAdapter.java [hunk 4]
```diff
@@ -86,6 +86,8 @@ private Schema read(JsonReader reader, Map<String, Schema> knownRecords) throws
     JsonToken token = reader.peek();
     switch (token) {
       case NULL:
+        // advance the value otherwise gson won't be able to get over the null value
+        reader.nextNull();
```
Comment (albertshau) [inline cdap-api-common/src/main/java/io/cdap/cdap/internal/io/SchemaTypeAdapter.java:90]: was this case never being hit before? Why is it being hit now?

Comment (yaojiefeng) [reply cdap-api-common/src/main/java/io/cdap/cdap/internal/io/SchemaTypeAdapter.java:90]: This is hit by https://github.com/cdapio/cdap/blob/develop/cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/spec/StageSpec.java#L46-L47 when I add the unit test for validation handler, basically if these two schema fields are null, I am able to do` GSON.toJson(stagespec)` but not able to transfer it back `GSON.fromJson(stagespecjson, StageSpec.class)` and gson complains it encountered a null while expecting a variable name.

File: cdap-api/src/main/java/io/cdap/cdap/api/plugin/PluginGroupConfig.java (hunk 1 above)
Comment (yaojiefeng) [reply cdap-api/src/main/java/io/cdap/cdap/api/plugin/PluginGroupConfig.java]: Initially the implementation is to have all the macro fields in the parent class to check, on a second thought to have it in the child class is better to implement a plugin, so removed this class and forced to use PluginConfig, it adds a bit complication on set the field for the nested config though

**Review by albertshau — APPROVED** [review]
